xml/rule-3.9.rng

```diff
@@ -6,73 +6,120 @@
+              <ref name="element-rule"/>
+
+              <!--
+               @COMPAT: The below expression types are invalid for element-rule
+               -->
+              <ref name="rsc_expression"/>
+              <ref name="op_expression"/>
+            </choice>
+          </oneOrMore>
+
+          <!--
+           @COMPAT: The below score attributes are invalid for element-rule
+           -->
           <optional>
             <choice>
               <externalRef href="score.rng"/>
               <attribute name="score-attribute"><text/></attribute>
             </choice>
           </optional>
-          <optional>
-            <attribute name="boolean-op">
-              <choice>
-                <value>or</value>
-                <value>and</value>
-              </choice>
-            </attribute>
-          </optional>
-          <optional>
-            <attribute name="role"><text/></attribute>
-          </optional>
+        </group>
+      </choice>
+    </element>
+  </define>
+
+  <!-- A rule element in a location constraint -->
+  <define name="element-rule-location">
+    <element name="rule">
+      <choice>
+        <attribute name="id-ref"><data type="IDREF"/></attribute>
+        <group>
+          <ref name="rule-common"/>
           <oneOrMore>
             <choice>
-              <ref name="expression"/>
+              <ref name="expression-location"/>
               <ref name="date_expression"/>
+              <ref name="element-rule-location"/>
+
+              <!--
+               @COMPAT: The below expression types are invalid for
+               element-rule-location
+               -->
               <ref name="rsc_expression"/>
               <ref name="op_expression"/>
-              <ref name="element-rule"/>
             </choice>
           </oneOrMore>
+          <optional>
+            <choice>
+              <externalRef href="score.rng"/>
+              <attribute name="score-attribute"><text/></attribute>
+            </choice>
+          </optional>
         </group>
       </choice>
     </element>
   </define>
 
-  <!-- A node attribute expression -->
-  <define name="expression">
-    <element name="expression">
-      <attribute name="id"><data type="ID"/></attribute>
-      <attribute name="attribute"><text/></attribute>
-      <attribute name="operation">
+  <!-- Attributes that are common to all rule elements -->
+  <define name="rule-common">
+    <attribute name="id"><data type="ID"/></attribute>
+    <optional>
+      <attribute name="boolean-op">
         <choice>
-          <value>lt</value>
-          <value>gt</value>
-          <value>lte</value>
-          <value>gte</value>
-          <value>eq</value>
-          <value>ne</value>
-          <value>defined</value>
-          <value>not_defined</value>
+          <value>or</value>
+          <value>and</value>
         </choice>
       </attribute>
+    </optional>
+    <optional>
+      <attribute name="role"><text/></attribute>
```

Comment:
Does role apply in all contexts? It wasn't mentioned in any of the special cases in T196.

Reply:
Yes